File: app/src/main/java/com/example/bandiaudiocapture/MainActivity.java
package com.example.bandiaudiocapture;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.media.projection.MediaProjectionManager;
import android.os.Bundle;
import android.os.Handler;
import android.view.Gravity;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

public class MainActivity extends Activity {

    private static final int REQ_AUDIO = 100;
    private static final int REQ_CAPTURE = 200;

    private TextView status;
    private final Handler handler = new Handler();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        LinearLayout layout = new LinearLayout(this);
        layout.setOrientation(LinearLayout.VERTICAL);
        layout.setPadding(50,100,50,50);
        layout.setGravity(Gravity.CENTER_HORIZONTAL);

        TextView title = new TextView(this);
        title.setText("반디 학습 녹음 테스트 v3");
        title.setTextSize(24);

        TextView guide = new TextView(this);
        guide.setText(
            "\n반디의 내부 음성을 WAV 파일로 저장합니다.\n" +
            "블루투스 이어폰을 그대로 사용하셔도 됩니다.\n"
        );
        guide.setTextSize(16);

        status = new TextView(this);
        status.setText("\n준비 완료\n");
        status.setTextSize(18);

        Button start = new Button(this);
        start.setText("학습 녹음 시작");
        start.setOnClickListener(v -> beginCapture());

        Button stop = new Button(this);
        stop.setText("학습 녹음 종료");
        stop.setOnClickListener(v -> {
            stopService(new Intent(this, CaptureService.class));
            status.setText("녹음을 종료하고 WAV 파일을 완성하는 중입니다...");
        });

        layout.addView(title);
        layout.addView(guide);
        layout.addView(status);
        layout.addView(start);
        layout.addView(stop);

        setContentView(layout);

        handler.post(updateStatus);
    }

    private final Runnable updateStatus = new Runnable() {
        @Override
        public void run() {

            boolean active = getSharedPreferences("capture", MODE_PRIVATE)
                    .getBoolean("active", false);

            float rms = getSharedPreferences("capture", MODE_PRIVATE)
                    .getFloat("rms", 0f);

            String file = getSharedPreferences("capture", MODE_PRIVATE)
                    .getString("file", "");

            long bytes = getSharedPreferences("capture", MODE_PRIVATE)
                    .getLong("bytes", 0);

            long seconds = getSharedPreferences("capture", MODE_PRIVATE)
                    .getLong("seconds", 0);

            boolean completed = getSharedPreferences("capture", MODE_PRIVATE)
                    .getBoolean("completed", false);

            if (active) {

                status.setText(
                    "● 녹음 중\n\n" +
                    "RMS: " + String.format("%.1f", rms) +
                    "\n녹음 시간: " + seconds + "초" +
                    "\n\n반디로 이동해서 학습하세요."
                );

            } else if (completed) {

                status.setText(
                    "✓ 녹음 완료\n\n" +
                    "녹음 시간: " + seconds + "초\n" +
                    "파일 크기: " + (bytes / 1024) + " KB\n\n" +
                    "저장 위치:\n" + file
                );
            }

            handler.postDelayed(this,500);
        }
    };

    private void beginCapture() {

        if (checkSelfPermission(Manifest.permission.RECORD_AUDIO)
                != PackageManager.PERMISSION_GRANTED) {

            requestPermissions(
                    new String[]{Manifest.permission.RECORD_AUDIO},
                    REQ_AUDIO
            );
            return;
        }

        getSharedPreferences("capture", MODE_PRIVATE)
                .edit()
                .putBoolean("completed", false)
                .apply();

        MediaProjectionManager manager =
                (MediaProjectionManager)
                        getSystemService(Context.MEDIA_PROJECTION_SERVICE);

        startActivityForResult(
                manager.createScreenCaptureIntent(),
                REQ_CAPTURE
        );
    }

    @Override
    public void onRequestPermissionsResult(
            int requestCode,
            String[] permissions,
            int[] grantResults) {

        super.onRequestPermissionsResult(
                requestCode, permissions, grantResults
        );

        if (requestCode == REQ_AUDIO
                && grantResults.length > 0
                && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            beginCapture();
        }
    }

    @Override
    protected void onActivityResult(
            int requestCode,
            int resultCode,
            Intent data) {

        super.onActivityResult(requestCode,resultCode,data);

        if (requestCode == REQ_CAPTURE
                && resultCode == RESULT_OK
                && data != null) {

            Intent service =
                    new Intent(this,CaptureService.class);

            service.putExtra("resultCode",resultCode);
            service.putExtra("data",data);

            startForegroundService(service);

            status.setText(
                    "녹음을 시작합니다...\n\n" +
                    "반디로 이동해서 영어 음성을 재생하세요."
            );

        } else if (requestCode == REQ_CAPTURE) {
            status.setText("화면 공유 권한이 허용되지 않았습니다.");
        }
    }

    @Override
    protected void onDestroy() {
        handler.removeCallbacks(updateStatus);
        super.onDestroy();
    }
}

File: app/src/main/java/com/example/bandiaudiocapture/MainActivity.kt
package com.example.bandiaudiocapture

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.media.*
import android.media.projection.MediaProjectionManager
import android.os.Bundle
import android.widget.*
import androidx.core.app.ActivityCompat
import kotlin.concurrent.thread
import kotlin.math.sqrt

class MainActivity : Activity() {

    private lateinit var status: TextView
    private var recorder: AudioRecord? = null
    @Volatile private var running = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val layout = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(40,80,40,40)
        }

        val title = TextView(this).apply {
            text = "반디 내부 오디오 테스트"
            textSize = 24f
        }

        status = TextView(this).apply {
            text = "\n아직 테스트하지 않았습니다.\n"
            textSize = 18f
        }

        val button = Button(this).apply {
            text = "내부 오디오 캡처 시작"
            setOnClickListener { startTest() }
        }

        layout.addView(title)
        layout.addView(status)
        layout.addView(button)

        setContentView(layout)
    }

    private fun startTest() {

        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.RECORD_AUDIO
            ) != PackageManager.PERMISSION_GRANTED) {

            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.RECORD_AUDIO),
                100
            )

            status.text = "마이크 권한을 허용한 뒤 버튼을 다시 눌러주세요."
            return
        }

        val manager =
            getSystemService(Context.MEDIA_PROJECTION_SERVICE)
                    as MediaProjectionManager

        startActivityForResult(
            manager.createScreenCaptureIntent(),
            200
        )
    }

    override fun onActivityResult(
        requestCode: Int,
        resultCode: Int,
        data: android.content.Intent?
    ) {
        super.onActivityResult(requestCode,resultCode,data)

        if (requestCode != 200 ||
            resultCode != RESULT_OK ||
            data == null) {
            status.text = "캡처 권한이 허용되지 않았습니다."
            return
        }

        val manager =
            getSystemService(Context.MEDIA_PROJECTION_SERVICE)
                    as MediaProjectionManager

        val projection =
            manager.getMediaProjection(resultCode,data)

        val config =
            AudioPlaybackCaptureConfiguration.Builder(projection)
                .addMatchingUsage(AudioAttributes.USAGE_MEDIA)
                .build()

        val format =
            AudioFormat.Builder()
                .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                .setSampleRate(44100)
                .setChannelMask(AudioFormat.CHANNEL_IN_STEREO)
                .build()

        val min =
            AudioRecord.getMinBufferSize(
                44100,
                AudioFormat.CHANNEL_IN_STEREO,
                AudioFormat.ENCODING_PCM_16BIT
            )

        recorder =
            AudioRecord.Builder()
                .setAudioFormat(format)
                .setBufferSizeInBytes(min * 2)
                .setAudioPlaybackCaptureConfig(config)
                .build()

        running = true
        recorder!!.startRecording()

        status.text =
            "캡처 중입니다.\n\n반디로 이동해서 영어를 재생하세요."

        thread {

            val buffer = ShortArray(min)

            while (running) {

                val n =
                    recorder?.read(buffer,0,buffer.size) ?: 0

                if (n > 0) {

                    var sum = 0.0

                    for (i in 0 until n) {
                        val v = buffer[i].toDouble()
                        sum += v*v
                    }

                    val rms = sqrt(sum/n)

                    runOnUiThread {

                        if (rms > 30) {
                            status.text =
                                "✓ 내부 오디오 감지됨!\n\nRMS: %.1f\n\n반디 일반 앱 캡처 가능성이 높습니다."
                                    .format(rms)
                        } else {
                            status.text =
                                "캡처 중...\nRMS: %.1f\n\n반디에서 영어를 재생해 주세요."
                                    .format(rms)
                        }
                    }
                }
            }
        }
    }
}
